web/resources/lib/component-loader/component-loader.ts

```diff
@@ -0,0 +1,263 @@
+
+    this.loadComponent()
+  }
+
+  disconnectedCallback() {
+    if (this.rendered && this.type === "svelte" && this.component) {
+      // prettier-ignore
+      try { this.component.$destroy() } catch {}
+    }
+
+    this.rendered = false
+    this.type = null
+    this.component = null
+  }
+}
+
+declare global {
+  interface HTMLElementTagNameMap {
+    "wj-component-loader": ComponentLoaderElement
+  }
+
+  interface Window {
+    ComponentLoaderElement: typeof ComponentLoaderElement
+  }
+}
+
+addElement(ComponentLoaderElement, "ComponentLoaderElement")
+
+// utility functions for getting this all to work
+
+/**
+ * Checks if the given class extends from `HTMLElement`.
+ *
+ * @param component - The class to check.
+ */
+function isHTMLComponent(component: any): component is typeof HTMLElement {
+  return component.prototype instanceof HTMLElement
+}
+
+/**
+ * Takes an `Element` and decontructs everything inside of it into a
+ * `DocumentFragment`, then returns the fragment. All children of the given
+ * element will be removed from the element.
+ *
+ * Taken from `svelte-tag`.
+ *
+ * @param from - The element to be decontructed.
+ */
+function unwrap(from: Element) {
+  let node = document.createDocumentFragment()
+  while (from.firstChild) {
+    node.appendChild(from.removeChild(from.firstChild))
+  }
+  return node
+}
+
+/**
+ * Gets all of the slots inside of this element, destructively.
+ *
+ * Taken from `svelte-tag`.
+ *
+ * @param from - The element to deconstruct the slots from.
+ */
+function takeSlots(from: Element) {
+  const namedSlots = from.querySelectorAll("[slot]")
+
+  let slots: Record<string, DocumentFragment> = {}
+
+  namedSlots.forEach(n => {
+    slots[n.slot] = unwrap(n)
+    from.removeChild(n)
+  })
+
+  // anything left over is the default slot
+  if (from.innerHTML.length) {
+    slots.default = unwrap(from)
+    from.innerHTML = ""
+  }
+
+  return slots
+}
+
+/**
+ * Creates a Svelte compatible slot handler object. Witchcraft!
+ *
+ * Taken from `svelte-tag`.
+ *
+ * @param slots - The slots to be used.
+ */
+function createSvelteSlotHandler(slots: Record<string, DocumentFragment>) {
+  const svelteSlots: Record<string, [ReturnType<typeof createSvelteSlotFunction>]> = {}
+
+  for (const slotName in slots) {
+    svelteSlots[slotName] = [createSvelteSlotFunction(slots[slotName])]
+  }
+
+  return svelteSlots
+}
+
+/**
+ * Creates a Svelte compatible slot function. Witchcraft!
+ *
+ * Taken from `svelte-tag`.
+ *
+ * @param fragment - The document fragment to be used as the slot.
+ */
+function createSvelteSlotFunction(fragment: DocumentFragment) {
+  return function () {
+    return {
+      c: noop,
+      m: function mount(target: Node, anchor?: Node) {
+        insert(target, fragment.cloneNode(true), anchor)
+      },
+      d: function destroy(detaching: boolean) {
+        // @ts-ignore
+        if (detaching && fragment.innerHTML) {
+          detach(fragment)
+        }
+      },
+      l: noop
```

Comment:
What are these single letters about? Svelte aliases?

Reply:
Yeah, I think they're just single letter names to make the component smaller when generated. This is internal Svelte witchcraft so I dunno exactly what everything does, lmao